### PR TITLE
Add IterativeSolverMonitor [feature/monitor]

### PR DESCRIPTION
**Pull request by jandrej**
This adds a new class `IterativeSolverMonitor` that can be attached to an iterative solver to monitor the residual and or the solution.

Please give your feedback on the design and implementation. Also consider adding the appropriate callbacks in other solver routines like `FGMRES::Mult` etc.

If the reviewers are OK with the concept, we can start with the changes in other routines than CG.

TODO:
- [x] CG
- [x] GMRES
- [x] FGMRES
- [x] NewtonSolver
- [x] ?

Disclaimer: This is almost an exact copy of the PETSc monitor.

Ping @barker29 @v-dobrev @jamiebramwell 

 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge| 
 | --- | --- | --- | --- | --- | --- | --- | 
| [#1390](https://github.com/mfem/mfem/pull/1390) | @jandrej | @tzanio | @jamiebramwell + @barker29 | 04/06/20 | 04/21/20 | ⌛due 04/28/20 | |

### CODE REVIEW

**Comment by barker29** [comment]
I have some ideas for simplification in #1393 let me know what you think.

**Comment by jandrej** [comment]
> I have some ideas for simplification in #1393 let me know what you think.

Looks better in every aspect. I'm going to merge this in here.

**Comment by jandrej** [comment]
I merged @barker29's changes and since there were no other suggestions/discussions for a few days now, I think this is ready for review.

File: examples/ex14p.cpp [hunk 1]
```diff
@@ -35,6 +35,38 @@
 using namespace std;
 using namespace mfem;
 
+class CustomSolverMonitor : public IterativeSolverMonitor
```
Comment (barker29) [inline examples/ex14p.cpp:38]: I suggest a more descriptive name, maybe `VisualizeSolutionMonitor` or `VisualizeIterateMonitor`.

**Comment by barker29** [comment]
I did an example monitor in `ex14p` that is part of the reason I'm interested in this PR. To do it I also had to add a `final` argument to the monitor, which I am not sure you actually want @jandrej . So feel free to roll back some or all of my commits.

**Comment by tzanio** [comment]
@jamiebramwell can you please review?

File: examples/ex10p.cpp [hunk 2]
```diff
@@ -45,6 +45,40 @@ using namespace mfem;
 
 class ReducedSystemOperator;
 
+class GeneralResidualMonitor : public IterativeSolverMonitor
```
Comment (jamiebramwell) [inline examples/ex10p.cpp:48]: It might be nice to add this simple monitor in regular src since it would be widely applicable to user problems.

**Comment by jandrej** [comment]
LGTM

**Comment by tzanio** [comment]
Please document in `CHANGELOG`, otherwise looks great!

**Comment by tzanio** [comment]
Merged in `next` for testing...

**Comment by v-dobrev** [comment]
From the internal tests, it looks like the monitors in ex10p (and probably ex14p) print from every rank, and in some cases they print when they are not supposed to output anything.

I think it will be best to try to keep the output from the examples the same as in master.

**Comment by v-dobrev** [comment]
> From the internal tests, it looks like the monitors in ex10p (and probably ex14p) print from every rank, and in some cases they print when they are not supposed to output anything.
> 
> I think it will be best to try to keep the output from the examples the same as in master.

@adrienbernede, @acfisher, do you know why this was not caught by the gitlab ci?

For example, compare the last gitlab ci: https://lc.llnl.gov/gitlab/mfem/mfem/pipelines/10372 with the output here: https://mybitbucket.llnl.gov/projects/MFEM/repos/autotest/browse/tux375/2020-04-23-llnl-gh-next/baseline-tux375.diff

**Comment by adrienbernede** [comment]
When you look at the baseline test on Gitlab, there is something wrong but it’s hard to tell what.
https://lc.llnl.gov/gitlab/mfem/mfem/-/jobs/68817

I agree that this kind of error should be made more visible. The deal was that baseline, if ending "with warnings", should be checked.

I will try to find a way to separate the comparison status from the execution status. But in the meantime, there is definitely something to investigate in this PR.

**Comment by v-dobrev** [comment]
Yes, the gitlab pipeline shows a problem (now resolved) but that is different from the issue I'm referring to, as seen in the above mybitbucket link. You are probably right that the issue maybe in the way the results from running `baseline` are processed.

**Comment by adrienbernede** [comment]
LGTM

**Comment by barker29** [comment]
`ex10p` should have better output in parallel now. My intention was specifically to change the output in this example, to demonstrate something simple you can do with a basic `IterativeSolverMonitor` (print info from each timestep, nonlinear step, and linear solve in a nicely indented readable way, which was basically impossible before). But if we want the output to match the previous example we can certainly do that.

**Comment by adrienbernede** [comment]
@v-dobrev I don’t see how the Gitlab issue was resolved?
@barker29 I is a good idea to have a variety of tests and outputs. It simply requires some work to update the baseline!

**Comment by adrienbernede** [comment]
LGTM

**Comment by vladotomov** [comment]
Sorry to intervene, I hope it's not too late to make a suggestion about this.

This kind of functionality can also be achieved with a slightly different approach, which we do in `NewtonSolver` - see the method `ProcessNewState`. People call this pattern a [template method](http://www.dofactory.com/net/template-method-design-pattern).

I think it's more flexible, as it exposes all fields of the `Solver` class, and the user can use anything he wants to track, without being restricted by the interface of `IterativeSolverMonitor`. The difference with the current code would be very small - the user derives the `Solver` and implements one method, instead of defining an `IterativeSolverMonitor`. The template method can be `const` to avoid changing the data of the base class.

**Comment by v-dobrev** [comment]
> @v-dobrev I don’t see how the Gitlab issue was resolved?

@adrienbernede, I meant it is resolved in `master` -- we can merge `master` here to get rid of the warnings.

**Comment by v-dobrev** [comment]
> `ex10p` should have better output in parallel now. My intention was specifically to change the output in this example, to demonstrate something simple you can do with a basic `IterativeSolverMonitor` (print info from each timestep, nonlinear step, and linear solve in a nicely indented readable way, which was basically impossible before). But if we want the output to match the previous example we can certainly do that.

@barker29, I agree, demonstrating this option is good but example 10/10p produces a lot of output already, so another example may be more appropriate. Maybe example 19p (added by @jamiebramwell) -- it has nonlinear solve but it is steady-state.

**Comment by v-dobrev** [comment]
Re-merged in `next`.

**Comment by v-dobrev** [comment]
LGTM

**Comment by v-dobrev** [comment]
> Sorry to intervene, I hope it's not too late to make a suggestion about this.
> 
> This kind of functionality can also be achieved with a slightly different approach, which we do in `NewtonSolver` - see the method `ProcessNewState`. People call this pattern a [template method](http://www.dofactory.com/net/template-method-design-pattern).
> 
> I think it's more flexible, as it exposes all fields of the `Solver` class, and the user can use anything he wants to track, without being restricted by the interface of `IterativeSolverMonitor`. The difference with the current code would be very small - the user derives the `Solver` and implements one method, instead of defining an `IterativeSolverMonitor`. The template method can be `const` to avoid changing the data of the base class.

I think hearing different perspectives is good. The template method approach has some advantages as you said (we may need to add a few parameters to the method for greater flexibility, e.g. current norm, current residual, etc). The fact that it is specific to a class can be an advantage, but also disadvantage. For example, the same monitor class can be re-used with different iterative solvers while with the template method approach you need to derive separate classes. For greater flexibility, we can support both approaches -- I don't see a big drawback in that.

There's yet another use for similar call-backs -- to allow users to control the algorithm in some way, e.g. to support custom convergence criteria, or custom line-search scaling as done with `NewtonSolver::ComputeScalingFactor`. These algorithm customizations can also be done with either template methods, or with calls to separate "monitor" classes.

**Comment by v-dobrev** [comment]
Regarding the gitlab ci: it looks like the warnings during the build step were preventing the rest of the test from being run.

In the last gitlab run I see all the differenes with the baseline. We need to update the baseline there with the latest changes in master -- right now there a lot of differences that are not from this branch.

**Comment by v-dobrev** [comment]
@barker29, @jandrej, the internal tests show that the modifications in `ex10p` produce a lot of extra output. I'd prefer if we demonstrate the new capability in an example that does not produce this much output, e.g. see https://github.com/mfem/mfem/pull/1390#issuecomment-619325748.

**Comment by v-dobrev** [comment]
Re-merged in `next`.

**Pull request merged**